sky/backends/backend_utils.py

```diff
@@ -1775,32 +1777,66 @@ def _update_cluster_status_no_lock(
         if external_ips is None or len(external_ips) == 0:
             raise exceptions.FetchIPError(
                 reason=exceptions.FetchIPError.Reason.HEAD)
-        if handle.launched_nodes == 1:
-            # Check the ray cluster status. We have to check it for single node
-            # case, since the get_node_ips() does not require ray cluster to be
-            # running.
-            ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml)
-            runner = command_runner.SSHCommandRunner(external_ips[0],
-                                                     **ssh_credentials)
-            returncode = runner.run('ray status', stream_logs=False)
-            if returncode:
-                raise exceptions.FetchIPError(
-                    reason=exceptions.FetchIPError.Reason.HEAD)
-        # If we get node ips correctly, the cluster is UP. It is safe to
-        # set the status to UP, as the `handle.external_ips` function uses ray
-        # to fetch IPs and starting ray is the final step of sky launch.
+        # Check if ray cluster status is healthy.
+        ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml)
+        runner = command_runner.SSHCommandRunner(external_ips[0],
+                                                 **ssh_credentials)
+        rc, output, _ = runner.run('ray status',
+                                   stream_logs=False,
+                                   require_outputs=True,
+                                   separate_stderr=True)
+        if rc:
+            raise exceptions.FetchIPError(
+                reason=exceptions.FetchIPError.Reason.HEAD)
+
+        def get_ready_nodes(pattern, output):
+            result = pattern.findall(output)
+            if len(result) == 0:
+                return 0
+            assert len(result) == 1, result
+            return int(result[0])
+
+        ready_workers = get_ready_nodes(_LAUNCHED_WORKER_PATTERN, output)
+        ready_head = get_ready_nodes(_LAUNCHED_HEAD_PATTERN, output)
+        assert ready_head <= 1, f'#head node should be <=1 (Got {ready_head}).'
+
+        if ready_head + ready_workers == handle.launched_nodes:
+            ray_cluster_up = True
```

Comment:
instead of directly copying the code here. can we refactor it out as a function so it can be reused in both places?

Reply:
I actually refactor a bit as they differ in some places but yeah I can try taking it out as a function.

Reply:
I just refactored a bit. I tried to incorporate `runner` into the function but looks like the output from `ray status` will be used here. so I ended up not taking `runner` but the output as the argument. PTAL, thanks!
https://github.com/skypilot-org/skypilot/blob/8628a254892114fdc4068d234976de731612c4b8/sky/backends/backend_utils.py#L1110